### PR TITLE
Mocking the system clock

### PR DESCRIPTION
**Pull request by yordanmadzhunkov**
I accidentally invoke everybody on the review. Feel free to leave, if you feel there are enough colleagues already invoked. For those who want to participate, I will explain the motivation:

We want to test our code using fuzz tests, because it provides benefits, using like random input 
 * to maximize code coverage
 * identify bugs in the code and recording the randomized input, so later a developer can replay/inspect that specific test
 
 In order to get those benefits, we need repeatable tests with require high level of determinism. Putting the `Utc::now` in blockchain header and computing hash is not deterministic. There are other source of not-deterministic behaviour like threads, random seed which will not be addressed in this review.
 
The goal of this PR is to make it possible to inject values into system clock before invoking the production code. The actual code in production should have no idea that we are mocking the system clock. 

Mocking the system clock.

### CODE REVIEW

**Comment by matklad** [comment]
Oh, fascinating, we are just having an interesting discussion of what time we should use in the first place at https://github.com/near/nearcore/pull/5097 cc @pmnoxx

File: core/primitives/src/time.rs
Comment (matklad) [inline core/primitives/src/time.rs:100]: Feels like significant amount of stuff in this file doesn't need to be `pub`. Right now, it's pretty hard to see what is the public interface I should use from other crates, and what is a private impl detail.

File: core/primitives/src/time.rs
Comment (matklad) [inline core/primitives/src/time.rs:136]: This sholdn't be called `system_time`, as https://doc.rust-lang.org/stable/std/time/struct.SystemTime.html is a very specific, different thing.

File: core/primitives/src/time.rs
Comment (yordanmadzhunkov) [reply core/primitives/src/time.rs]: Done, thanks for the suggestion

**Review by bowenwang1996 — COMMENTED** [review]
Please change the title of your PR to a descriptive one. Thanks!

File: core/primitives/src/time.rs
Comment (yordanmadzhunkov) [reply core/primitives/src/time.rs:95]: Awesome suggestion. Everything becomes much simpler.

File: core/primitives/src/time.rs
Comment (yordanmadzhunkov) [reply core/primitives/src/time.rs]: Thanks

**Comment by yordanmadzhunkov** [comment]
@matklad 
I also have mixed feelings. 
Specific design issues with the current solution:

1. You understood it correctly. Refactoring the example code to call now once, I can break the tests, (only if you check number of times timer is called) 

2. I will implement
`"if time is mocked { get mocked instance or panic } else { return real time}"`

3. Good point. Looks like the real clock ::now method also has build in mutex. At this point I will not be worried about performance issues

4. I am also not sure, what is the right approach

5. The motivation to make time mocking local to specific thread is the fact that test run in parallel in separate threads. If time mocking is not thread specific, different tests interfere between each other.

**Comment by yordanmadzhunkov** [comment]
@matklad 
Actually, on point 4. I think the correct approach is to create a clock interface and use dependency injection. This way, we can mock tests and in production we can simply inject real system clock. This will require a lot of work to pass the instance to the clock between functions, classes and etc.

**Comment by pmnoxx** [comment]
I like the idea of adding mocking to time. While writing tests I had to add a bit of code to support testing code, which gets executed rarely, having mocks for time would be useful.

However:
- A test usually spawns multiple threads. Most likely you will have to change code that starts all actors in the code, in such a way, that child thread uses the same time mock as it's parent.
I would use something like:
```
thread_local! {
    pub static Singletor: Optional<Arc<RefCell<MockClockPerThread>>>> = ...;
};
```
Everytime you start new actor, yo would make it point to the same Arc<RefCell<MockClockPerThread>>>` as it parent.

- We also don't mock time for Actix, timeouts, etc. It would be great to be able to add an option to increase speed of progression of the time to test timeouts, etc.

- I'm starting to think that it may be a good idea refactor our code to use one library for time. That would make code much cleaner, we wouldn't have to mock different type of times. 
I would be in favor of having our own clock.

```
struct Clock { 
    // number of ns since 1970 / utc time
    ns: u64
}
struct Duration { 
    // difference in ns
    ns: i64
}
```
function:
- now()
- duration_since(...)
- sub(...) // singled difference
- we can add `BorshSerialize` / `BorshDeserialize` / this can be used to replace `u64` as time storage.

This would allow us to use single source of time for all code, and also it would be easier to mock.
Sometimes we use `u64`, sometimes `Instant::now`, `Datetime<Utc>::new()`, etc.

**Review by bowenwang1996 — COMMENTED** [review]
@yordanmadzhunkov after a discussion with @matklad we agree that the design in this implementation is not optimal. However, to not get blocked on this for fuzzing, I suggest that you figure out the minimal amount of change we need to make the fuzzing possible (deterministic). There are likely a few places where we can refactor the functions to have time passed in as a parameter so that we can avoid doing a big change like this at once.

File: core/primitives/src/time.rs
Comment (yordanmadzhunkov) [reply core/primitives/src/time.rs]: Done

File: runtime/runtime-params-estimator/src/main.rs
Comment (matklad) [inline runtime/runtime-params-estimator/src/main.rs:192]: Here we always want to use the real clock.

File: runtime/runtime-params-estimator/src/testbed_runners.rs
Comment (matklad) [inline runtime/runtime-params-estimator/src/testbed_runners.rs:7]: ditto here: basically, all the code in `runtime-params-estimator` should be exempt from time mocking. This is essentially benchmarking code, and it has its' own requrenments for time, unrelated to the rest of the code.

File: core/primitives/src/time.rs
Comment (matklad) [inline core/primitives/src/time.rs:67]:
```suggestion
    pub fn get() -> MutexGuard<'static, Clock> {
        SINGLETON.lock().unwrap()
    }
```

would simplify call-sites.

File: core/primitives/src/time.rs
Comment (matklad) [inline core/primitives/src/time.rs:58]: Rather than maintaining per-thread map, Arc and lazy ourselves, we can leverage std::therad_local:

```rust
struct MockClockPerThread {
    utc: VecDeque<DateTime<Utc>>,
    durations: VecDeque<Duration>,
    utc_call_count: u64,
    instant_call_count: u64,
    instant: Instant,
    is_mock: bool,
}

impl MockClockPerThread {
    fn with<F, T>(f: F) -> T
    where
        F: FnOnce(&mut MockClockPerThread) -> T,
    {
        thread_local! {
            static INSTANCE: RefCell<MockClockPerThread> = RefCell::default()
        }
        INSTANCE.with(|it| f(&mut *it.borrow_mut()))
    }

    fn pop_utc(&mut self) -> Option<DateTime<chrono::Utc>> {
        self.utc_call_count += 1;
        self.utc.pop_front()
    }
}

pub struct Clock {}

impl Clock {
    pub fn add_utc(mock_date: DateTime<chrono::Utc>) {
        MockClockPerThread::with(|clock| {
            clock.utc.push_back(mock_date);
            clock.is_mock = true;
        });
    }
    pub fn utc() -> DateTime<chrono::Utc> {
        MockClockPerThread::with(|clock| {
            let x = clock.pop_utc();
            match x {
                Some(t) => t,
                None => {
                    if clock.is_mock {
                        panic!("Mock clock run out of samples");
                    }
                    chrono::Utc::now()
                }
            }
        })
    }
}
```

File: core/primitives/src/time.rs
Comment (matklad) [inline core/primitives/src/time.rs:74]: :thinking: I am surprised that this passes the tests. If two tests are run on the same thread, the first test uses mocking, but the second doesn't, the second should fail at runtime, as it'll hit `is_mock = true` from the first test. For the "thread-local singleton" design, I'd probably expect some kind of RAII API here: 

```rust
#[test] 
fn test_with_mock_time() {
  let mut tm = Clock::time_machine(); // sets `is_mock` to  true
  tm.add_utc("July 3, 1985");

  // implimit `drop(tm)` sets `is_mock` to false;
}
```

File: core/primitives/Cargo.toml
Comment (matklad) [inline core/primitives/Cargo.toml:33]: We no longer need this dependency.

File: core/primitives/src/time.rs
Comment (matklad) [inline core/primitives/src/time.rs:15]: No need to keep this `pub`. It's relatively important to not make things `pub`, as knowing what API exends past the current compilation unit makes reading the code easier.

**Review by matklad — APPROVED** [review]
Approving to unblock, but there are three fixes we need to do here: 

* remove unneded cargo dependency 
* remove changes to the estimator
* make sure that `is_mock` can't leak between tests

The `thread_local` suggestion is optional -- I do like to minimize the amount of machinery we use, but as this is impl details, this doesnt' really matter. 

Two other nice-to-haves:

* are we sure that this change is enough to enable the fuzzing we want? Given that we are not mocking tokio's time or `thread::sleep`s I wouldn't be surprised if this isn't actually enough.
* are we sure that this is the minimal change? From my conversation with @bowenwang1996, I recall the idea was to just refactor some specific part of the code-base to take time as an argument, without introducing thread-locals, but also without fixing *every* usage of time. If that turns out to be infeasible, I think it's OK to do what we are doing here, but, still, let's try to minimize the blast radius of the design which we agree is not the long-term one.

File: test-utils/loadtester/src/stats.rs
Comment (bowenwang1996) [inline test-utils/loadtester/src/stats.rs:88]: It seems that we can rewrite this as 
```rust
while curr_height <= self.to_height.unwrap() {
   ...
}
```

File: test-utils/loadtester/src/stats.rs
Comment (bowenwang1996) [inline test-utils/loadtester/src/stats.rs:1]: I believe `test-utils/loadtester` is removed in master. What is the reason for restoring it?

**Review by bowenwang1996 — APPROVED** [review]
Approve to unblock, but please address Aleksey's comments.

File: core/primitives/src/time.rs
Comment (matklad) [reply core/primitives/src/time.rs:58]: To clarify -- the current design also works I think, no need to rewrite if you don't like the new version better.

Comment (matklad) [reply core/primitives/src/time.rs:58]: :thinking: that being said, I think there's a potential problem with the hash map -- we never remove entries form it. If we do maintain hashmap in the prod code, and we do create new threads from time to time, than this would leak memory. If that is indeed the problem (which I dont' know), we should switch to `thread_local`, as that includes the necessary code to release resources when a thread exists.

File: chain/network/src/routing.rs
Comment (matklad) [inline chain/network/src/routing.rs:830]: These `&mut Clock {}` feel odd, perhaps we should keep `::instant` to be an associated fn (without self)?

**Comment by pmnoxx** [comment]
I created my own implementation Clock for purpose of mocking time. Please, consider an alternative:
https://github.com/near/nearcore/pull/5175

File: core/primitives/src/time.rs
Comment (pmnoxx) [inline core/primitives/src/time.rs:85]: How about we change signature to `pub fn utc_now() -> DateTime<chrono::Utc>`

- self is not used - This will allow us to use `Clock::utc_now()` without having to create `Clock {}`
- changing `utc` to `utc_now` should make code more readable. Without prior knowledge we will know that we are getting current time.

**Review by pmnoxx — APPROVED** [review]
LGTM

**Comment by pmnoxx** [comment]
@yordanmadzhunkov I pushed a few huge changes to `chain/network`, you may have to resolve merge conflicts.

Good job on the PR.
I look forward to seeing the next PR, where you add propagation of `Clock` to throughout our code base.
After that we can work on cleaning up the code, to make sure we only have one structure representing time instead of 4.

**Comment by pmnoxx** [comment]
@chefsale @frol Can you take a look?

File: core/primitives/src/time.rs
Comment (yordanmadzhunkov) [reply core/primitives/src/time.rs]: Done

File: core/primitives/src/time.rs
Comment (yordanmadzhunkov) [reply core/primitives/src/time.rs]: Done :)

File: chain/network/src/routing.rs
Comment (yordanmadzhunkov) [reply chain/network/src/routing.rs]: Done

File: test-utils/loadtester/src/stats.rs
Comment (yordanmadzhunkov) [reply test-utils/loadtester/src/stats.rs:88]: Removed, it was added by accident

File: core/primitives/src/time.rs
Comment (yordanmadzhunkov) [reply core/primitives/src/time.rs]: This will be reworked later.

File: core/primitives/src/time.rs
Comment (yordanmadzhunkov) [reply core/primitives/src/time.rs]: yes, you are right :)

File: test-utils/loadtester/src/stats.rs
Comment (yordanmadzhunkov) [reply test-utils/loadtester/src/stats.rs]: no reason, it looks like it happened by accident

**Review by matklad — APPROVED** [review]
:+1: 

Couple of more nitpicly coments:

* some `pub` can still be removed
* Rather than making the user import `MockClockGuard`, the API could look like this:

```rust
let _guard: MockClockGuard = Clock::mock();
```

:thinking: thinking more about this, I think it even makes sense to make `add_utc` and such methods of the guard object, rather than of the `Clock` itself. 

Anyway, I don't think we should try to polish Rust API in this PR (we must polish it sometime later though, time mocking is a fundamental fascility used thourgouht the codebase, its importatn that the API is just right, otherwise changing it later would be a pain).

**Review by mzhangmzz — APPROVED** [review]

**Pull request merged**
